Add per-fragment information to OpenGL Pixel History

zm2he opened the pull request:
<!--
Before submitting a pull request you are strongly recommended to read the
docs/CONTRIBUTING.md file which gives some information on how to prepare a
change:

https://github.com/baldurk/renderdoc/blob/v1.x/docs/CONTRIBUTING.md

For small changes you don't have to read the document end to end, but should at
least look at the sections on how to ensure your code and commits are formatted
according to the style requirements.
-->

## Description
- Modified fragment counting method to use stencil counting to be more accurate
- Add a per-fragment callback function to calculate postmod primitive id and shader output
- To calculate primitive id, we replace the fragment shader with the primitive id fragment shader in this program
- Add a function to mark failed depth tests for fragments

<!--
Describe here what your pull request changes and why it should happen. For small
changes which are obvious this can just be a line or two - even the commit
message is sometimes enough.
-->

TonyTascioglu committed:
Add per-fragment information to OpenGL Pixel History

orsonbraines commented:
@baldurk Some of the Linux compilation CI tasks are failing to install compilation dependencies. Do you know why that would be happening?

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -52,6 +57,82 @@ enum class OpenGLTest
+  GLuint *attachedShaders = new GLuint[numAttachedShaders];
+
+  driver->glGetAttachedShaders(currentProgram, numAttachedShaders, &numAttachedShaders,
+                               attachedShaders);
```
Please don't use glGetAttachedShaders like this, this isn't reliable in the general case of GL captures. Look at how `GLReplay::RenderOverlay` works to take the vertex-stage shaders from an existing program and link them with a custom fragment shader. That will also handle program pipelines.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -52,6 +57,82 @@ enum class OpenGLTest
+  driver->FillReflectionArray(ProgramRes(driver->GetCtx(), newProgram), dstStages);
+
+  PerStageReflections stages;
+  driver->FillReflectionArray(ProgramRes(driver->GetCtx(), newProgram), stages);
```
The reflection array for the source stages should be filled from the source program, not the new program.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -75,6 +156,16 @@ bool PixelHistorySetupResources(WrappedOpenGL *driver, GLPixelHistoryResources &
+  // This is because it requires scalars to be swizzled something enabled
+  // in version 420. We also require intBitsToFloat which is introduced in
+  // version 330.
+  int GLSLVersion = 420;
```
Since the primary audience of GL use is on old/outdated environments we should try as much as possible to support older GL versions. In this case the swizzling can be removed, and versions before 330 can be supported by writing the primtiive ID as a float directly which can uniquely represent primitives up to 16777215 which should typically be enough for those old GL users. It would be good to use 330 if available for `intBitsToFloat` but it's not a hard requirement.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -97,7 +194,7 @@ rdcarray<EventUsage> QueryModifyingEvents(WrappedOpenGL *driver, GLPixelHistoryR
   driver->ReplayLog(0, events[0].eventId, eReplay_WithoutDraw);
   // execute the occlusion queries
-  for(size_t i = 0; i < events.size(); i++)
+  for(size_t i = 0; i < events.size(); ++i)
```
I don't really mind which convention you use from a stylistic point of view, but please don't make changes like this on existing code as it only adds noise into the git history and diffs. Note that `++i` is never necessary in this context since there is nothing to be ordered against for prefix/postfix.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -135,9 +232,10 @@ rdcarray<EventUsage> QueryModifyingEvents(WrappedOpenGL *driver, GLPixelHistoryR
-      driver->glGetQueryObjectiv(occlusionQueries[i], eGL_QUERY_RESULT, &numFragments);
-      if(numFragments > 0)
+      uint32_t numSamples;
+      driver->glGetQueryObjectuiv(occlusionQueries[i], eGL_QUERY_RESULT, &numSamples);
```
What's the reason for the variable rename here? Again either name would be fine but the change generates undesirable noise unless there's a specific reason.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -152,42 +250,51 @@ rdcarray<EventUsage> QueryModifyingEvents(WrappedOpenGL *driver, GLPixelHistoryR
+      driver->glClearStencil(0);
+      driver->glClear(eGL_STENCIL_BUFFER_BIT);
+      driver->glEnable(eGL_STENCIL_TEST);
+      driver->ReplayLog(modEvents[i].eventId, modEvents[i].eventId, eReplay_Full);
```
You should not use `eReplay_Full` here if you're only replaying a single draw, you should use `eReplay_OnlyDraw`.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -276,13 +383,226 @@ void QueryFailedTests(WrappedOpenGL *driver, GLPixelHistoryResources &resources,
+                            const std::map<uint32_t, uint32_t> &eventFragments,
+                            PerFragmentQueryType queryType)
+{
+  driver->ReplayLog(0, modEvent.eventId - 1, eReplay_Full);
```
This should replay up to `modEvent.eventId` with `eReplay_WithoutDraw` if you're intending to set up the state to patch before then replaying the draw.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -276,13 +383,226 @@ void QueryFailedTests(WrappedOpenGL *driver, GLPixelHistoryResources &resources,
+
+    GLint currentProgram;
+
+    driver->glGetIntegerv(eGL_CURRENT_PROGRAM, &currentProgram);
```
This program isn't restored afterwards. Currently it doesn't matter that the state is polluted but this can easily cause problems in future if the state should remain undisturbed after the work.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -276,13 +383,226 @@ void QueryFailedTests(WrappedOpenGL *driver, GLPixelHistoryResources &resources,
+}
+
+void WriteClearValues(WrappedOpenGL *driver, GLPixelHistoryResources &resources,
+                      const EventUsage clearEvent, int x, int y, rdcarray<PixelModification> &history)
```
Why do you need a separate path for fetching postmod values for clears? Ideally it should be possible to fetch all postmods (except for postmods between different fragments in the same event) in a single pass while calculating per-event overdraw counts. This also means that postmods are fetched at the correct precision, since we want to correctly represent limited precision depending on the target format.

orsonbraines replied:
Do we care about reading back stencil values for postmod? Otherwise I agree that we should move this all in a single post mod pass.

baldurk replied:
For after each _event_ then yes, we want stencil values and should be able to obtain them along with depth and colour when doing a single pass (not changing the draws at all). Stencil values between different fragments of a draw we can't get with the current system because we're using stencil for fragment counting, that's a limitation we have to accept.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -276,13 +383,226 @@ void QueryFailedTests(WrappedOpenGL *driver, GLPixelHistoryResources &resources,
+    case eGL_GEQUAL: return shaderOutputDepth >= depthInBuffer;
+    case eGL_ALWAYS: return true;
+
+    default: RDCASSERT(0);
```
You should never assert on false/0. If there is an error condition please add an `RDCERR` with a meaningful error message like `unexpected depth function`.

baldurk reviewed (changes requested):
I've left some specific notes, but this seems to be moving in the wrong direction with regard to the number of repeated replays of the whole capture. It was my mistake not to call this out more explicitly in previous PRs but I assumed it was a temporary development stage and that the passes would be refactored as more was added in.

Generally speaking replaying the capture is the most expensive operation we can do so it should be limited as much as possible. I know you are likely using the D3D11 implementation as reference, and it is currently very inefficient in the number of times it will replay the whole capture - I think I mentioned this in initial discussions elsewhere. The vulkan pixel history is a much better model in terms of the number of passes it does. Partly this is enforced by the command-buffer recording nature of vulkan, and enabled by the ability to patch out shader writes, but generally you should be aiming to only do a small number of whole-capture replays and gather the information you need as you go, and at all costs avoid doing a full replay on every event or every modification.

You should aim at least to not do any more than D3D11, and wherever possible combine passes together.

orsonbraines commented:
Thanks for the quick feedback! We'll look over it all and address what we can in a couple days.

zm2he committed:
Improved fragment id shader code

- Deleted WriteClearValues and replace with QueryPostModValues
- Add support for earlier GLSL versions and fix logic for creating primitive id program
- Changed to only run per fragment queries on events with multiple fragments

orsonbraines commented:
> I've left some specific notes, but this seems to be moving in the wrong direction with regard to the number of repeated replays of the whole capture. It was my mistake not to call this out more explicitly in previous PRs but I assumed it was a temporary development stage and that the passes would be refactored as more was added in.
> 
> Generally speaking replaying the capture is the most expensive operation we can do so it should be limited as much as possible. I know you are likely using the D3D11 implementation as reference, and it is currently very inefficient in the number of times it will replay the whole capture - I think I mentioned this in initial discussions elsewhere. The vulkan pixel history is a much better model in terms of the number of passes it does. Partly this is enforced by the command-buffer recording nature of vulkan, and enabled by the ability to patch out shader writes, but generally you should be aiming to only do a small number of whole-capture replays and gather the information you need as you go, and at all costs avoid doing a full replay on every event or every modification.
> 
> You should aim at least to not do any more than D3D11, and wherever possible combine passes together.

We've tried to use the Vulkan implementation as a reference, but none of us have much experience with Vulkan and so it's hard for us to read and understand. Right now, I see that the way we are using QueryPerFragmentValues is really inefficient, so we'll restructure it so it does not need to be called in a loop. We can also make it so that we count the number of fragments and get the postmod values in the same pass.

For any other performance improvements we need to make, we'd probably need more specific pointers.

baldurk commented:
No worries at all, I don't expect perfection and especially since the D3D11 pixel history is already doing many replays it doesn't have to be optimal. It's just something I wanted to raise as there's currently more replays than necessary in the way you're writing it. As a rule of thumb a big enough capture could be 0.5-1s in the worst case for a single replay so you want to keep the number down as much as possible.

The main aim should be to gather as much as possible into a single replay pass where possible, inserting extra calls as you go as needed. Anything that doesn't require replaying an action multiple times is generally a good candidate here, though there's a bit of handwavey flexibility we take advantage of in terms of side-effects and dependencies between actions that we assume will remain mostly valid.

For pixel history we should be able to at least grab premod and postmod values for each event, as well as the occlusion count to know if the event touched our pixel at all, all in a single pass. Currently the D3D11 pixel history will then do one replay per candidate, per test that's both enabled & might fail. Then it does one final combined replay to fetch the per-fragment shader output values.

Playing it safe by replaying once per-action to avoid side-effect issues is fine and can be improved later should it prove to be a problem (as the D3D11 one does). In principle this could probably be improved but it becomes complex in the general case. It's currently not handled robustly, but it mostly works for practical applications since most people don't tend to write side-effects in such a way that they can become a problem like this.

orsonbraines commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -52,9 +58,222 @@ enum class OpenGLTest
+  PrimitiveId
+};
+
+GLuint CreatePrimitiveIdProgram(GLuint Program, GLuint Pipeline, GLuint fragShader,
```
To fix the way we replaced fragment shaders fro primitive ID, we copied the code from GLReplay::CreateOverlayProgram and fixed it up to work for pixel history. Should we try to refactor this into a common function we can use in both places?

baldurk replied:
Yes I think this function is probably complex enough that it merits being made into a helper, probably in `GLReplay` itself. It looks like if you refactored it to do the delete/create of the new program outside then the function is already fairly factored out.

orsonbraines commented:
We're still working on more changes to reduce the number of replays when querying the per fragment values, but we pushed our other changes in the meantime.

baldurk commented on renderdoc/data/glsl/pixelhistory_primid_legacy.frag:
```diff
@@ -0,0 +1,32 @@
+
+void main(void)
+{
+  color_out = vec4(float(gl_PrimitiveID));
```
Rather than creating a new shader for this I think it would be better to use the same shader with an `#ifdef`. Since the difference is quite minor this makes it easier for future maintenance if anything needs to change (as unlikely as that is) and makes it easier to read and realise that there are two minor variants on the same shader.

In theory we could use the builtin preprocessor defines for language version, but given that this is GL I don't trust that to work reliably so it'd be better to use our own define.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -250,9 +398,48 @@ rdcarray<EventUsage> QueryModifyingEvents(WrappedOpenGL *driver, GLPixelHistoryR
+    driver->glGetIntegerv(eGL_READ_FRAMEBUFFER_BINDING, &savedReadFramebuffer);
+    // bind our own framebuffer to save the pixel values
+    driver->glBindFramebuffer(eGL_FRAMEBUFFER, resources.frameBuffer);
+    driver->ReplayLog(modEvents[i].eventId, modEvents[i].eventId, eReplay_Full);
```
Using `eReplay_Full` here is not the intended use for replaying a single draw. In practice currently if you specify the same EIDs for start and end it will be equivalent, but generally `eReplay_Full` and `eReplay_WithoutDraw` are reserved for replaying ranges of events (either an half-closed or entirely closed interval, effectively), with `eReplay_OnlyDraw` for replaying a single draw.

I admit this is confusing, especially when there's redundancy between the options and which EIDs you specify - because this API is internal it's been unchanged pretty much since the first implementation. It might merit a refactor in future to be more clear but that's not a priority at the moment.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -250,9 +398,48 @@ rdcarray<EventUsage> QueryModifyingEvents(WrappedOpenGL *driver, GLPixelHistoryR
+
+    // restore the capture's framebuffer
+    driver->glBindFramebuffer(eGL_DRAW_FRAMEBUFFER, savedDrawFramebuffer);
+    driver->glBindFramebuffer(eGL_READ_FRAMEBUFFER, savedReadFramebuffer);
```
There's an issue here - when replaying the capture in this pass we're not actually writing any of the draws to the original framebuffers. This means that subsequent passes depending on the results of earlier passes may be completely wrong because the data isn't there - worst case it could lead to a GPU crash if the data affects control flow.

It also has a minor problem here that we're fetching the postmod data at 'full' precision because our offscreen framebuffer is full floats. Part of the value of the postmod data is that it's at 'real' precision - e.g. if the texture is RGBA8 then this shows any quantisation that happens from the shader output value.

I think a better solution here is for each draw to replay it as normal, but then use a single pixel framebuffer blit to copy the output (after quantisation) to our framebuffer. I would also recommend that you blit each pixel to a new slot in the framebuffer and then do all the reads at the end - as this stands at the moment it's syncing the GPU every draw to readback the results, but you can do all of this work pipelined.

baldurk reviewed (changes requested):
I think this is a good direction, I've left some notes on things I saw doing a first pass. I will do a more in-depth review once all of the changes are ready.

orsonbraines committed:
Improve GL Pixel History per PR feedback

- Blit fragments and read back results in batch
- Modify per fragment callbacks to reduce number of full replays
- Use defines in the primitiveId shader to toggle between float/int
  output

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -277,13 +593,390 @@ void QueryFailedTests(WrappedOpenGL *driver, GLPixelHistoryResources &resources,
+    driver->glGetIntegerv(eGL_DRAW_FRAMEBUFFER_BINDING, &savedDrawFramebuffer);
+    driver->glGetIntegerv(eGL_READ_FRAMEBUFFER_BINDING, &savedReadFramebuffer);
+
+    for(size_t j = 0; j < std::max(numFragments, 1u); ++j)
```
Please use `RDCMAX` rather than `std::max`.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -52,10 +65,118 @@ enum class OpenGLTest
+    }
+    if(curStencil != 0)
+    {
+      driver->CreateTextureImage(resources.stencilImage, depthFormat, eGL_NONE, eGL_NONE,
```
Should be `stencilFormat` I think here.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -157,28 +335,129 @@ void QueryPostModPixelValues(WrappedOpenGL *driver, GLPixelHistoryResources &res
+        eGL_COLOR_BUFFER_BIT |
+            (resources.depthTextureAttachedToFrameBuffer ? eGL_DEPTH_BUFFER_BIT : eGL_NONE) |
+            (resources.stencilTextureAttachedToFrameBuffer ? eGL_STENCIL_BUFFER_BIT : eGL_NONE),
+        eGL_NEAREST);
```
This doesn't get the shader output values because it's reading from the target framebuffer, meaning the output has already been quantised by whatever format there is there. To get the shader output values we need to render offscreen to RGBA32F to be sure we get the correct full precision values.

baldurk commented on renderdoc/driver/gl/gl_pixelhistory.cpp:
```diff
@@ -277,13 +593,390 @@ void QueryFailedTests(WrappedOpenGL *driver, GLPixelHistoryResources &resources,
+      driver->glStencilFunc(eGL_EQUAL, (int)j, 0xff);
+      driver->glClear(eGL_STENCIL_BUFFER_BIT);
+
+      driver->ReplayLog(modEvents[i].eventId, modEvents[i].eventId, eReplay_OnlyDraw);
```
This has the same kind of problem as above - shader output values must be fetched direct to RGBA32F, otherwise if the original capture's color target is RGBA8 then we can't tell the difference between an output of 0.999 and 0.998 because they will be rounded to 1.0.

Note this is in contrast to premod/postmod which are _deliberately_ at the target output precision. That way you can see what value was output and then what value was actually stored to the target.

baldurk reviewed (changes requested):
This is looking good now, though I've left some comments about issues remaining.

orsonbraines committed:
Use full precision texture format for shaderOut

orsonbraines committed:
Bind framebuffer before clearing stencil buffer
